lib/location/method_gnss.c

```diff
@@ -0,0 +1,357 @@
+#include <zephyr.h>
+#include <logging/log.h>
+
+#include <modem/location.h>
+#include <modem/lte_lc.h>
+#include <modem/at_cmd.h>
+#include <nrf_modem_gnss.h>
+#include <nrf_errno.h>
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+#include <net/nrf_cloud_rest.h>
+#include <net/nrf_cloud_agps.h>
+#endif
+
+#include "loc_core.h"
+
+LOG_MODULE_DECLARE(location, CONFIG_LOCATION_LOG_LEVEL);
+
+/* range 10240-3456000000 ms, see AT command %XMODEMSLEEP */
+#define MIN_SLEEP_DURATION_FOR_STARTING_GNSS 10240
+#define AT_MDM_SLEEP_NOTIF_START "AT%%XMODEMSLEEP=1,%d,%d"
+#define AT_MDM_SLEEP_NOTIF_STOP "AT%XMODEMSLEEP=0" /* not used at the moment */
+#define AGPS_REQUEST_RECV_BUF_SIZE 3500
+#define AGPS_REQUEST_HTTPS_RESP_HEADER_SIZE 500
+
+extern location_event_handler_t event_handler;
+extern struct loc_event_data current_event_data;
+
+struct k_work_args {
+	struct k_work work_item;
+	struct loc_gnss_config gnss_config;
+};
+
+static struct k_work_args method_gnss_start_work;
+struct k_work method_gnss_fix_work;
+struct k_work method_gnss_timeout_work;
+struct k_work method_gnss_agps_request_work;
+
+static int fix_attemps_remaining;
+static bool first_fix_obtained;
+static bool running;
+
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+static char rest_api_recv_buf[CONFIG_NRF_CLOUD_REST_FRAGMENT_SIZE + AGPS_REQUEST_HTTPS_RESP_HEADER_SIZE];
+static char agps_data_buf[AGPS_REQUEST_RECV_BUF_SIZE];
+#endif
+
+static K_SEM_DEFINE(entered_psm_mode, 0, 1);
+
+void method_gnss_lte_ind_handler(const struct lte_lc_evt *const evt)
+{
+	switch (evt->type) {
+	case LTE_LC_EVT_MODEM_SLEEP_ENTER:
+		if (evt->modem_sleep.type == LTE_LC_MODEM_SLEEP_PSM) {
+			/* Signal to method_gnss_positioning_work_fn() that GNSS can be started. */
+			k_sem_give(&entered_psm_mode);
+		}
+		break;
+
+	case LTE_LC_EVT_MODEM_SLEEP_EXIT:
+		/* Prevent GNSS from starting while LTE is running. */
+		k_sem_reset(&entered_psm_mode);
+		break;
+	default:
+		break;
+	}
+}
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+static void method_gnss_agps_request_work_fn(struct k_work *item)
+{
+	struct nrf_cloud_rest_context rest_ctx = {
+		.connect_socket = -1,
+		.keep_alive = false,
+		.timeout_ms = CONFIG_NRF_CLOUD_REST_RECV_TIMEOUT * MSEC_PER_SEC,
+		.auth = CONFIG_AGPS_LOCATION_SERVICE_NRF_CLOUD_JWT_STRING,
+		.rx_buf = rest_api_recv_buf,
+		.rx_buf_len = sizeof(rest_api_recv_buf),
+		.fragment_size = 0
+	};
+
+	struct nrf_cloud_rest_agps_request request =
+		{ NRF_CLOUD_REST_AGPS_REQ_ASSISTANCE, NULL, NULL};
+
+	struct nrf_cloud_rest_agps_result result = {agps_data_buf, sizeof(agps_data_buf), 0};
+
+	nrf_cloud_rest_agps_data_get(&rest_ctx, &request, &result);
+
+	nrf_cloud_agps_process(result.buf, result.agps_sz, NULL);
+}
+#endif
+
+void method_gnss_event_handler(int event)
+{
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_FIX:
+		LOG_DBG("GNSS: Got fix");
+		k_work_submit_to_queue(loc_core_work_queue_get(), &method_gnss_fix_work);
+		break;
+
+	case NRF_MODEM_GNSS_EVT_PVT:
+		LOG_DBG("GNSS: Got PVT event");
```

Comment:
We could log some information about satellistes here. At minimum number of found satellites. If easily doable, also their IDs. But let's keep the log in oneline so the log won't flood with too much information.

Reply:
Here's one example from asset_tracker_v2: https://github.com/nrfconnect/sdk-nrf/blob/6d79b4833d69e0fb864a1a1c3f083adc4e51d3bf/applications/asset_tracker_v2/src/modules/gps_module.c#L403

It prints a one liner for each tracked satellite, so it probably prints too much information. But at least the code shows how to iterate through all satellites from the PVT notification.

Reply:
@hiltunent or @tokangas Please try to get this done